plugins/org.eclipse.xtext/src/org/eclipse/xtext/serializer/sequencer/ContextFinder.java

```diff
@@ -91,160 +92,188 @@
+		List<IContext> candidates = fromEObjects(contextCandidates, semanticObject);
+		return fromIContexts(findByContents(semanticObject, candidates));
+	}
+
+	@Override
+	@Deprecated
+	public Iterable<EObject> findContextsByContentsAndContainer(EObject semanticObject,
 			Iterable<EObject> contextCandidates) {
-		Map<IConstraint, List<EObject>> result = Maps.newLinkedHashMap();
-		for (EObject ctx : contextCandidates) {
-			IConstraint constraint = constraints.get(Tuples.create(ctx, semanticObject.eClass()));
-			if (ctx == null)
-				continue;
-			List<EObject> ctxs = result.get(constraint);
-			if (ctxs == null)
-				result.put(constraint, ctxs = Lists.newArrayList());
-			ctxs.add(ctx);
+		List<IContext> candidates = fromEObjects(contextCandidates, semanticObject);
+		return fromIContexts(findByContentsAndContainer(semanticObject, candidates));
+	}
+
+	protected Multimap<IConstraint, IContext> getConstraints(EObject sem) {
+		EClass type = sem == null ? null : sem.eClass();
+		Multimap<IConstraint, IContext> result = ArrayListMultimap.create();
+		for (Entry<IContext, IConstraint> e : constraints.entrySet()) {
+			IContext context = e.getKey();
+			IConstraint constraint = e.getValue();
+			if (constraint.getType() == type) {
+				result.put(constraint, context);
+			}
+		}
+		return result;
+	}
+
+	protected Multimap<IConstraint, IContext> getConstraints(EObject sem, Iterable<IContext> contextCandidates) {
+		EClass type = sem == null ? null : sem.eClass();
+		Multimap<IConstraint, IContext> result = ArrayListMultimap.create();
+		for (IContext ctx : contextCandidates) {
+			IConstraint constraint = constraints.get(ctx);
+			if (constraint != null && constraint.getType() == type) {
+				result.put(constraint, ctx);
+			}
 		}
 		return result;
 	}
 
-	protected EObject getRootContext() {
+	protected IContext getRootContext(EObject sem) {
 		for (AbstractRule rule : ruleNames.getAllRules())
 			if (GrammarUtil.isEObjectRule(rule))
-				return rule;
+				return SerializationContext.fromEObject(rule, sem);
 		throw new RuntimeException("There is no parser rule in the grammar.");
 	}
 
 	protected void initConstraints() {
-		if (constraintContexts == null) {
+		if (constraints == null) {
 			constraints = Maps.newLinkedHashMap();
-			constraintContexts = grammarConstraintProvider.getConstraints(ruleNames.getContextGrammar());
-			//			System.out.println(Joiner.on("\n").join(constraintContexts));
-			for (IConstraintContext ctx : constraintContexts)
-				for (IConstraint constraint : ctx.getConstraints())
-					constraints.put(Tuples.create(ctx.getContext(), constraint.getType()), constraint);
+			constraints = grammarConstraintProvider.getConstraints(ruleNames.getContextGrammar());
 		}
 	}
 
-	protected boolean intersect(Iterable<EObject> it1, Iterable<EObject> it2) {
-		for (EObject i1 : it1)
-			for (EObject i2 : it2)
-				if (i1 == i2)
+	protected boolean intersect(Iterable<IContext> it1, Iterable<IContext> it2) {
+		for (IContext i1 : it1)
```

Comment:
I don't know about the avg number of elements in the iterables, but it could be more efficient to use CollectionExtensions.toSet(it1) and than a single loop over second one with set.contains(elem)